# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

treeducken <- function(params_file) {
    .Call(`_treeducken_treeducken`, params_file)
}

#' Simulates species tree using constant rate birth-death process
#'
#' @details At present only can simulate to a number of tips. This function does
#' so using the general algorithm of Hartmann et al. 2010.
#' @param sbr_ species birth rate (i.e. speciation rate)
#' @param sdr_ species death rate (i.e. extinction rate)
#' @param numbsim_ number of species trees to simulate
#' @param n_tips_ number of tips to simulate to
#' @return List of objects of the tree class (as implemented in APE)
#' @references
#' K. Hartmann, D. Wong, T. Stadler. Sampling trees from evolutionary models.
#'     Syst. Biol., 59(4): 465-476, 2010.
#'
#' T. Stadler. Simulating trees on a fixed number of extant species.
#'     Syst. Biol., 60: 676-684, 2011.
#' @examples
#' mu <- 0.5 # death rate
#' lambda <- 2.0 # birth rate
#' numb_replicates <- 10
#' numb_extant_tips <- 4
#' # simulate trees under the GSA so first simulates a tree with
#' # numb_extant_tips * 100 tips counting each time we have a tree with 10 tips
#' # then randomly picks one of those trees
#'
#' sim_sptree_bdp(sbr_ = lambda,
#'                 sdr_ = mu,
#'                 numbsim_ = numb_replicates,
#'                 n_tips_ = numb_extant_tips)
sim_sptree_bdp <- function(sbr_, sdr_, numbsim_, n_tips_) {
    .Call(`_treeducken_sim_sptree_bdp`, sbr_, sdr_, numbsim_, n_tips_)
}

#' Simulates locus tree using constant rate birth-death-transfer process
#'
#' @details Given a species tree simulates a locus or gene family tree along
#'     the species tree.
#' @param species_tree_ species tree to simulate along
#' @param gbr_ gene birth rate
#' @param gdr_ gene death rate
#' @param lgtr_ gene trasnfer rate
#' @param num_loci_ number of locus trees to simulate
#' @return List of objects of the tree class (as implemented in APE)
#' @references
#' Rasmussen MD, Kellis M. Unified modeling of gene duplication, loss, and
#'     coalescence using a locus tree. Genome Res. 2012;22(4):755â€“765.
#'     doi:10.1101/gr.123901.111
#' @examples
#' # first simulate a species tree
#' mu <- 0.5 # death rate
#' lambda <- 2.0 # birth rate
#' numb_replicates <- 10
#' numb_extant_tips <- 4
#' # simulate trees under the GSA so first simulates a tree with
#' # numb_extant_tips * 100 tips counting each time we have a tree with 10 tips
#' # then randomly picks one of those trees
#'
#' sp_tree <- sim_sptree_bdp(sbr_ = lambda,
#'                 sdr_ = mu,
#'                 numbsim_ = numb_replicates,
#'                 n_tips_ = numb_extant_tips)
#'
#' gene_br <- 1.0
#' gene_dr <- 0.2
#' transfer_rate <- 0.2
#' sim_locustree_bdp(species_tree_ = sp_tree,
#'                   gbr_ = gene_br,
#'                   gdr_ = gene_dr,
#'                   lgtr_ = transfer_rate,
#'                   num_loci_ = 10)
sim_locustree_bdp <- function(species_tree_, gbr_, gdr_, lgtr_, num_loci_) {
    .Call(`_treeducken_sim_locustree_bdp`, species_tree_, gbr_, gdr_, lgtr_, num_loci_)
}

#' Simulates a cophylogenetic system using a paired birth-death process
#'
#' @details Simulates a cophylogenetic system using birth-death processes. The
#'     host tree is simulated following a constant rate birth-death process
#'     with an additional parameter - the cospeciation rate. This rate works as
#'     the speciation rate with the additional effect that if cospeciation
#'     occurs the symbiont tree also speciates. The symbiont tree is related to
#'     the host tree via an association matrix that describes which lineages
#'     are associated with which. The symbiont tree has an independent
#'     birth-death process with the addition of a host shift speciation rate
#'     that allows for the addition of more associated hosts upon symbiont
#'     speciation.
#' @param species_tree_ species tree to simulate along
#' @param hbr_ host tree birth rate
#' @param hdr_ host tree death rate
#' @param sbr_ symbiont tree birth rate
#' @param sdr_ symbiont tree death rate
#' @param host_exp_rate_ host shift speciation rate
#' @param cosp_rate_ cospeciation rate
#' @param timeToSimTo_ time units to simulate until
#' @param numbsim_ number of replicates
#' @return A list containing the `host_tree`, the `symbiont_tree`, the
#'     association matrix at present, and the history of events that have
#'     occurred.
#' @examples
#' # first simulate a species tree
#' mu <- 0.5 # death rate
#' lambda <- 2.0 # birth rate
#' numb_replicates <- 10
#' numb_extant_tips <- 4
#' # simulate trees under the GSA so first simulates a tree with
#' # numb_extant_tips * 100 tips counting each time we have a tree with 10 tips
#' # then randomly picks one of those trees
#'
#' sp_tree <- sim_sptree_bdp(sbr_ = lambda,
#'                 sdr_ = mu,
#'                 numbsim_ = numb_replicates,
#'                 n_tips_ = numb_extant_tips)
#'
sim_cophylo_bdp <- function(hbr_, hdr_, sbr_, sdr_, host_exp_rate_, cosp_rate_, timeToSimTo_, numbsim_) {
    .Call(`_treeducken_sim_cophylo_bdp`, hbr_, hdr_, sbr_, sdr_, host_exp_rate_, cosp_rate_, timeToSimTo_, numbsim_)
}

